### PR TITLE
Initial commit of amp-access-scroll

### PR DESCRIPTION
**Pull request by kushal**
- Adds an access vendor named scroll
- If user is authenticated, adds UI element to page

### CODE REVIEW

**Comment by aghassemi** [comment]
@dvoytenko Anyone else familiar with access you recommend to review this?
/cc @rudygalfi

**Comment by dvoytenko** [comment]
@aghassemi I'm on it.

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * from ClientAdapter.
+ * @implements {../../amp-access/0.1/access-vendor.AccessVendor}
+ */
+export class ScrollAccessVendor extends AccessClientAdapter {
+  constructor(ampdoc, accessService) {
+    super(ampdoc, CONFIG, {
+      buildUrl: accessService.buildUrl.bind(accessService),
+      collectUrlVars: accessService.collectUrlVars_.bind(accessService),
+    });
+    this.element_ = null;
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: jsdoc

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            this.element_.show();
+          }
+        });
+  }
+}
+
+/**
+ * @extends {HTMLElement}
+ */
+class ScrollElement extends HTMLElement {
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Is this now possible on all browsers? I feel like I've run into some compatibility issues before.

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  constructor(ampdoc) {
+    super();
+    installStylesForDoc(ampdoc, CSS, () => {}, false, TAG);
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Could you pls check if `installStylesForDoc` protects from inserting stylesheet twice?

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    this.ampdoc = ampdoc;
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: jsdoc here and below.

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    this.wrapper_ = document.createElement('div');
+    this.wrapper_.classList.add('amp-access-scroll-bar');
+    this.appendChild(this.wrapper_);
+
+
+    this.placeholder_ = document.createElement('div');
+    this.placeholder_.classList.add('amp-access-scroll-placeholder');
+    this.wrapper_.appendChild(this.placeholder_);
+
+    const img = document.createElement('amp-img');
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Also, totally fine to use just normal ``, especially for a logo. This kind of logo might just be better to do via CSS background, but up to you.

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    img.setAttribute('height', 26);
+    this.placeholder_.appendChild(img);
+
+    this.iframe_ = document.createElement('iframe');
+    this.iframe_.setAttribute('scrolling', 'no');
+    this.iframe_.setAttribute('frameborder', '0');
+    this.iframe_.setAttribute('allowtransparency', 'true');
+    this.iframe_.setAttribute('title', 'Scroll');
+    this.iframe_.setAttribute('width', '100%');
+    this.iframe_.setAttribute('height', '100%');
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Is it possible to set `sandbox` for this iframe?

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    this.wrapper_.appendChild(this.iframe_);
+  }
+
+  show() {
+    Services.accessServiceForDoc(this.ampdoc)
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: `accessService` is already passed in the constructor. Can you save it there and use it here?

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        .then(accessService => accessService.getAccessReaderId())
+        .then(readerId => {
+          this.iframe_.onload = () => {
+            this.wrapper_.removeChild(this.placeholder_);
+          };
+          this.iframe_.setAttribute('src',
+              'https://connect.scroll.com/amp/scrollbar?readerId=' + readerId);
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Let's do `encodeURIComponent(readerId)`. Should be websafe, but this way it's for sure.

File: extensions/amp-access-scroll/0.1/scroll-impl.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        });
+  }
+}
+
+customElements.define('amp-access-scroll-elt',
```
Comment (dvoytenko) [inline extensions/amp-access-scroll/0.1/scroll-impl.js]: Do we have to define it in CE? That's a bit rife with incompatibilties. I think it'd be just safer all around to just have ScrollElement a wrapper/holder class. Is there a strong reason to register it in CE?

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 1 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: done

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 2 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: i think i was confused about what you wanted here, put something in for now

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 3 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: yup, it does

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 4 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: done

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 5 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: just changed to img for now

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 6 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: should be fine

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 8 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: done

File: extensions/amp-access-scroll/0.1/scroll-impl.js (hunk 9 above)
Comment (kushal) [reply extensions/amp-access-scroll/0.1/scroll-impl.js]: see above

**Comment by kushal** [comment]
By the way, I think the answer is no, but is there any way for our extension to also automagically provision our amp-analytics URLs?

**Comment by jpettitt** [comment]
@kushal please create a .md file with doc for your extension

**Pull request merged**